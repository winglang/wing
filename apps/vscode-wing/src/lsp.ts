import { ExtensionContext } from "vscode";
import { LanguageClientOptions } from "vscode-languageclient";
import {
  LanguageClient,
  Executable,
  ServerOptions,
} from "vscode-languageclient/node";
import { getWingBinAndArgs } from "./bin-helper";
import { LANGUAGE_SERVER_ID, LANGUAGE_SERVER_NAME } from "./constants";

export class LanguageServerManager {
  client: LanguageClient | undefined;

  constructor(public readonly context: ExtensionContext) {}

  async stop() {
    await this.client?.stop();
  }

  async start() {
    const args = await getWingBinAndArgs(this.context);

    if (!args) {
      // User doesn't have wing and doesn't want to install it yet
      return;
    }

    args.push("lsp");

    const run: Executable = {
      command: args[0]!,
      args: args.slice(1),
      options: {
        env: {
          ...process.env,
        },
      },
    };
    const serverOptions: ServerOptions = {
      run,
      debug: run,
    };
    let clientOptions: LanguageClientOptions = {
      documentSelector: [
        { scheme: "file", language: "wing", pattern: "**/*.w" },
      ],
    };

    // Create the language client and start the client.
    this.client = new LanguageClient(
      LANGUAGE_SERVER_ID,
      LANGUAGE_SERVER_NAME,
      serverOptions,
      clientOptions
    );

    await this.client.start();
  }
}
