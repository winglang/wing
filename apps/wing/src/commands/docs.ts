import open from "open";

export async function docs() {
  await open("https://docs.winglang.io");
}