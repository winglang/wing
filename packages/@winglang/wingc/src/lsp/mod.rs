mod code_actions;
mod completions;
mod document_symbols;
mod goto_definition;
mod hover;
mod rename_prepare;
mod rename_request;
mod rename_visitor;
mod signature;
mod symbol_locator;
mod sync;
