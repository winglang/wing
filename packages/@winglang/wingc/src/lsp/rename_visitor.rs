use lsp_types::{Position, PrepareRenameResponse, Range, TextEdit};

use crate::diagnostic::WingLocation;
use crate::type_check::symbol_env::{LookupResult, SymbolEnv};
use crate::type_check::{
	resolve_super_method, SymbolKind, Types, UnsafeRef, CLASS_INFLIGHT_INIT_NAME, CLASS_INIT_NAME,
};
use crate::visit::{visit_scope, Visit};
use crate::visit_context::{VisitContext, VisitorWithContext};
use crate::{ast::*, visit_context};
pub struct LinkedSymbol<'a> {
	symbol: Symbol,
	references: Vec<&'a Symbol>,
}
pub struct RenameVisitor<'a> {
	types: &'a Types,
	linked_symbols: Vec<LinkedSymbol<'a>>,
	ctx: VisitContext,
}

impl<'a> RenameVisitor<'a> {
	pub fn new(types: &'a Types) -> Self {
		Self {
			types,
			linked_symbols: vec![],
			ctx: VisitContext::new(),
		}
	}

	fn is_symbol_linked(&mut self, symbol: &'a Symbol) -> bool {
		self
			.linked_symbols
			.iter()
			.any(|s: &LinkedSymbol<'a>| symbol.same(&s.symbol) || s.references.iter().any(|r| symbol.same(r)))
	}

	fn add_reference_symbol(&mut self, symbol: &'a Symbol, symbol_env: Option<&UnsafeRef<SymbolEnv>>) {
		// symbols that appear in let/if lef statements will point to a prev declaration of a variable of the same name if exists
		// this is why we add them in advance during visit_statement
		// the other condition is for "this" that points to the "new" keyword for some reason
		if self.is_symbol_linked(symbol) || symbol.name == "this" {
			return;
		}
		if let Some(env) = symbol_env.or(self.ctx.current_env()) {
			match env.lookup_ext(symbol, None) {
				LookupResult::Found(symbol_kind, lookup_info) | LookupResult::NotPublic(symbol_kind, lookup_info) => {
					// TODO: remove to support rename-refactor of namespaces - after adjusting the edit
					if matches!(symbol_kind, &SymbolKind::Namespace(_)) {
						// as their rename include " as ___"
						return;
					}
					// Default spans are either generated or global, so we don't want to rename them
					if lookup_info.span.is_default() {
						return;
					}
					if let Some(linked) = self
						.linked_symbols
						.iter_mut()
						.find(|s| lookup_info.span.eq(&s.symbol.span))
					{
						linked.references.push(symbol);
					} else {
						self.linked_symbols.push(LinkedSymbol {
							symbol: Symbol {
								name: symbol.name.clone(),
								span: lookup_info.span.clone(),
							},
							references: if lookup_info.span.eq(&symbol.span) {
								vec![]
							} else {
								vec![symbol]
							},
						});
					}
				}
				_ => {}
			}
		}
	}

	pub fn create_text_edits(&mut self, position: Position, new_text: String) -> Vec<TextEdit> {
		let location = WingLocation {
			line: position.line,
			col: position.character,
		};
		for symbol in &mut self.linked_symbols {
			if symbol.symbol.span.contains_location(&location) {
				return format_references_to_edit(symbol, new_text);
			}

			let mut is_found = false;
			// to remove the lock we must get out of the for loop
			for child in symbol.references.iter_mut() {
				if child.span.contains_location(&location) {
					is_found = true;
					break;
				}
			}
			if is_found {
				return format_references_to_edit(symbol, new_text);
			}
		}
		vec![]
	}

	fn prepare_symbol_rename(&self, symbol: &Symbol) -> PrepareRenameResponse {
		return PrepareRenameResponse::RangeWithPlaceholder {
			range: Range {
				start: Position {
					line: symbol.span.start.line,
					character: symbol.span.start.col,
				},
				end: Position {
					line: symbol.span.end.line,
					character: symbol.span.end.col,
				},
			},
			placeholder: symbol.name.clone(),
		};
	}

	pub fn prepare_rename(&mut self, position: Position) -> PrepareRenameResponse {
		let location = WingLocation {
			line: position.line,
			col: position.character,
		};
		for symbol in &self.linked_symbols {
			if symbol.symbol.span.contains_location(&location) {
				return self.prepare_symbol_rename(&symbol.symbol);
			}

			// to remove the lock we must get out of the for loop
			for child in symbol.references.iter() {
				if child.span.contains_location(&location) {
					return self.prepare_symbol_rename(child);
				}
			}
		}
		PrepareRenameResponse::DefaultBehavior {
			default_behavior: false,
		}
	}
}

impl<'a> VisitorWithContext for RenameVisitor<'a> {
	fn ctx(&mut self) -> &mut visit_context::VisitContext {
		&mut self.ctx
	}
}

impl<'a> Visit<'a> for RenameVisitor<'a> {
	fn visit_scope(&mut self, node: &'a Scope) {
		self.ctx.push_env(self.types.get_scope_env(&node));
		visit_scope(self, node);
		self.ctx.pop_env();
	}

	fn visit_symbol(&mut self, node: &'a Symbol) {
		self.add_reference_symbol(node, None);
		crate::visit::visit_symbol(self, node);
	}

	fn visit_expr(&mut self, node: &'a Expr) {
		match &node.kind {
			ExprKind::JsonMapLiteral { fields, .. } => {
				let type_ = self.types.maybe_unwrap_inference(self.types.get_expr_type(node));
				let type_ = *if let Some(type_) = self.types.get_type_from_json_cast(node.id) {
					*type_
				} else {
					type_
				}
				.maybe_unwrap_option();

				if let Some(c) = type_.as_struct() {
					for (field, ..) in fields {
						self.add_reference_symbol(field, Some(&UnsafeRef::from(&c.env)));
					}
				}
			}
			ExprKind::Call { arg_list, callee } => {
				let Some(env) = self.ctx.current_env() else {
					// usually the env will exist, if not- we cannot use it to resolve the super method.
					// it's the same as used here: https://github.com/winglang/wing/blob/main/packages/@winglang/wingc/src/lsp/symbol_locator.rs#L374-L376
					return;
				};
				// we need to get the struct from the callee - to get the right env
				let callee_type = match callee {
					CalleeKind::Expr(expr) => self.types.get_expr_type(expr),
					CalleeKind::SuperCall(method) => resolve_super_method(method, &env, &self.types)
						.ok()
						.map_or(self.types.error(), |t| t.0),
				}
				.maybe_unwrap_option()
				.to_owned();

				if let Some(func) = callee_type.as_function_sig() {
					if let Some(arg) = func.parameters.last() {
						let struct_type = arg.typeref.maybe_unwrap_option().as_struct();
						if let Some(s) = struct_type {
							for (arg, ..) in &arg_list.named_args {
								self.add_reference_symbol(arg, Some(&UnsafeRef::from(&s.env)));
							}
						}
					}
				}
			}
			_ => {}
		}
		crate::visit::visit_expr(self, node);
	}

	fn visit_reference(&mut self, node: &'a Reference) {
		match node {
			Reference::InstanceMember { property, object, .. } => {
				let object_type = self.types.get_expr_id_type_ref(object.id);
				if let Some(inner_env) = object_type.as_env() {
					self.add_reference_symbol(property, Some(&UnsafeRef::from(inner_env)));
				}
			}
			_ => {}
		}

		crate::visit::visit_reference(self, node);
	}

	fn visit_stmt(&mut self, stmt: &'a Stmt) {
		match &stmt.kind {
			StmtKind::IfLet(IfLet { var_name, .. }) => self.linked_symbols.push(LinkedSymbol {
				symbol: var_name.clone(),
				references: vec![],
			}),
			StmtKind::Let { var_name, .. } => self.linked_symbols.push(LinkedSymbol {
				symbol: var_name.clone(),
				references: vec![],
			}),
			//TODO: to be handled in a following PR, renaming interface fields is not supported yet
			// StmtKind::Interface(c) => {
			// for field in &c.methods {
			// 	self.linked_symbols.push(LinkedSymbol {
			// 		symbol: field.0.clone(),
			// 		references: vec![],
			// 	})
			// }
			// }
			StmtKind::Struct(s) => {
				for field in &s.fields {
					self.linked_symbols.push(LinkedSymbol {
						symbol: field.name.clone(),
						references: vec![],
					})
				}
			}
			StmtKind::Class(c) => {
				for (m, ..) in &c.methods {
					if m.name == CLASS_INIT_NAME || m.name == CLASS_INFLIGHT_INIT_NAME {
						continue;
					}
					self.linked_symbols.push(LinkedSymbol {
						symbol: m.clone(),
						references: vec![],
					})
				}
				for f in &c.fields {
					self.linked_symbols.push(LinkedSymbol {
						symbol: f.name.clone(),
						references: vec![],
					})
				}
			}
			_ => {}
		}
		crate::visit::visit_stmt(self, stmt);
	}
}

fn format_references_to_edit(linked: &mut LinkedSymbol, new_text: String) -> Vec<TextEdit> {
	let mut edits = vec![];
	edits.push(TextEdit {
		new_text: String::from(&new_text),
		range: Range {
			start: Position {
				line: linked.symbol.span.start.line,
				character: linked.symbol.span.start.col,
			},
			end: Position {
				line: linked.symbol.span.end.line,
				character: linked.symbol.span.end.col,
			},
		},
	});

	for child in linked.references.iter_mut() {
		edits.push(TextEdit {
			new_text: String::from(&new_text),
			range: Range {
				start: Position {
					line: child.span.start.line,
					character: child.span.start.col,
				},
				end: Position {
					line: child.span.end.line,
					character: child.span.end.col,
				},
			},
		});
	}

	edits
}
