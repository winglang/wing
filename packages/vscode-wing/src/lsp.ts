import { workspace } from "vscode";
import { LanguageClientOptions } from "vscode-languageclient";
import {
  LanguageClient,
  Executable,
  ServerOptions,
} from "vscode-languageclient/node";
import { LANGUAGE_SERVER_ID, LANGUAGE_SERVER_NAME } from "./constants";

export class LanguageServerManager {
  client: LanguageClient | undefined;
  wingBin: string | undefined;

  setWingBin(wingBin: string) {
    this.wingBin = wingBin;
  }

  async stop() {
    await this.client?.stop();
    await this.client?.dispose();
    this.client = undefined;
  }

  async start() {
    await this.stop();
    if (!this.wingBin) {
      return;
    }

    const run: Executable = {
      command: this.wingBin,
      args: ["lsp", "--no-update-check"],
      options: {
        env: {
          ...process.env,
        },
      },
    };
    const serverOptions: ServerOptions = {
      run,
      debug: run,
    };
    let clientOptions: LanguageClientOptions = {
      initializationOptions: {
        parameterHints: workspace
          .getConfiguration("editor.parameterHints", {
            languageId: "wing",
          })
          .get<boolean>("enabled", true),
      },
      documentSelector: [
        { scheme: "file", language: "wing", pattern: "**/*.w" },
      ],
    };

    // Create the language client and start the client.
    this.client = new LanguageClient(
      LANGUAGE_SERVER_ID,
      LANGUAGE_SERVER_NAME,
      serverOptions,
      clientOptions
    );

    await this.client.start();
  }
}
