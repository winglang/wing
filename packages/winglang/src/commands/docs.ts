import open from "open";

export async function docs() {
  await open("https://www.winglang.io/docs/");
}
