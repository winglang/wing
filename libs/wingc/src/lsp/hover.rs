use crate::docs::Documented;
use crate::lsp::sync::PROJECT_DATA;
use crate::type_check::symbol_env::LookupResult;
use crate::visit::Visit;
use crate::wasm_util::extern_json_fn;
use lsp_types::{Hover, HoverContents, MarkupContent, MarkupKind};

use super::symbol_locator::{SymbolLocator, SymbolLocatorResult};
use super::sync::{check_utf8, WING_TYPES};

#[no_mangle]
pub unsafe extern "C" fn wingc_on_hover(ptr: u32, len: u32) -> u64 {
	extern_json_fn(ptr, len, on_hover)
}
pub fn on_hover(params: lsp_types::HoverParams) -> Option<Hover> {
	WING_TYPES.with(|types| {
		let types = types.borrow_mut();
		PROJECT_DATA.with(|project_data| {
			let project_data = project_data.borrow();
			let uri = params.text_document_position_params.text_document.uri.clone();
			let file = check_utf8(uri.to_file_path().expect("LSP only works on real filesystems"));
			let root_scope = &project_data.asts.get(&file).unwrap();

			let mut symbol_finder = SymbolLocator::new(&types, params.text_document_position_params.position.into());
			symbol_finder.visit_scope(root_scope);

			if let Some(lookup) = symbol_finder.lookup_located_symbol() {
				if let LookupResult::Found(symbol_info, ..) = &lookup {
					let docs = symbol_info.render_docs();
					let span = symbol_finder.located_span()?;
					return Some(Hover {
						contents: HoverContents::Markup(MarkupContent {
							kind: MarkupKind::Markdown,
							value: docs,
						}),
						range: Some(span.into()),
					});
				}
			} else {
				if let SymbolLocatorResult::LooseField { field_type, field, .. } = &symbol_finder.result {
					let span = symbol_finder.located_span().unwrap();
					let value = format!("```wing\n{}: {field_type}\n```", field.name);

					return Some(Hover {
						contents: HoverContents::Markup(MarkupContent {
							kind: MarkupKind::Markdown,
							value,
						}),
						range: Some(span.into()),
					});
				}
			}

			None
		})
	})
}

#[cfg(test)]
mod tests {
	use crate::lsp::hover::*;
	use crate::lsp::sync::test_utils::*;
	use lsp_types::*;

	/// Creates a snapshot test for a given wing program's hover at a given position
	/// In the wing program, place a comment "//^" into the text where the "^" is pointing to the desired character position
	///
	/// First parameter will be the name of the tests, as well as the identifier to use for the list of completion in the asserts (see last parameter)
	/// Second parameter is the wing code block as a string literal
	/// After the first two parameters, any additional are optional statements that should be used for asserting on the given hover data.
	///
	/// Result is a [Hover] object
	macro_rules! test_hover_list {
		($name:ident, $code:literal) => {
			test_hover_list!($name, $code,);
		};
		($name:ident, $code:literal, $($assertion:stmt)*) => {
			#[test]
			fn $name() {
				// NOTE: this is needed for debugging to work regardless of where you run the test
				std::env::set_current_dir(env!("CARGO_MANIFEST_DIR")).unwrap();

				let text_document_position_params = load_file_with_contents($code);
				let hover = on_hover(HoverParams {
					text_document_position_params,
					work_done_progress_params: Default::default(),
				});

				if let Some($name) = hover {
					insta::with_settings!(
						{
							prepend_module_to_snapshot => false,
							omit_expression => true,
							snapshot_path => "./snapshots/hovers",
						}, {
							insta::assert_yaml_snapshot!($name);
						}
					);
					$($assertion)*
				} else {
					panic!("Expected hover data");
				}
			}
		};
	}

	test_hover_list!(
		just_variable,
		r#"
		let myString = "hello";
        //^
		"#,
	);

	test_hover_list!(
		new_expression_nested,
		r#"
bring cloud;
new cloud. 
		//^
"#,
	);

	test_hover_list!(
		class_symbol,
		r#"
		bring cloud;

		let bucket = new cloud.Bucket();
        //^
"#,
	);

	test_hover_list!(
		class_symbol_in_closure,
		r#"
inflight class MyClass { }

inflight () => {
  let myClass = new MyClass();
    //^

}
"#,
	);

	test_hover_list!(
		class_property,
		r#"
bring cloud;

let bucket = new cloud.Bucket();
bucket.addObject
      //^
"#,
	);

	test_hover_list!(
		static_stdtype_method,
		r#"
Json.stringify(123);
      //^
"#,
	);

	test_hover_list!(
		inside_inflight_test,
		r#"
test "hello" {
  let a = 1;
    //^
}
"#,
	);

	test_hover_list!(
		inside_class_method,
		r#"
class Foo {
  bar() {
    let hello = 1;
        //^
  }
}
"#,
	);

	test_hover_list!(
		inside_class_field,
		r#"
bring cloud;
class Foo {
  my_bucket: cloud.Bucket;
    //^
}
"#,
	);

	test_hover_list!(
		new_statement,
		r#"
bring cloud;
new cloud.Bucket();
          //^
"#
	);

	test_hover_list!(
		user_defined_types,
		r#"
class Foo { }
     //^
"#
	);

	test_hover_list!(
		user_defined_type_annotation,
		r#"
class Foo { }
let a: Foo = new Foo();
      //^
"#
	);

	test_hover_list!(
		user_defined_type_reference_property,
		r#"
class Foo { 
	static static_method() { }
}
Foo.static_method();
   //^
"#
	);

	test_hover_list!(
		user_defined_type_reference_type,
		r#"
class Foo { 
	static static_method() { }
}
Foo.static_method();
//^
"#
	);

	test_hover_list!(
		static_method,
		r#"
class Foo {
  static my(a: str, b: bool): str { return "str"; }
}

Foo.my();
  //^
"#
	);

	test_hover_list!(
		builtin_in_preflight,
		r#"
assert(true);
//^
"#
	);

	test_hover_list!(
		builtin_in_inflight,
		r#"
class Foo {
  inflight bar() {
    assert(true);
    //^
  }
}
"#
	);

	test_hover_list!(
		test_bring_sdk,
		r#"
bring cloud;
      //^
"#
	);

	test_hover_list!(
		test_var,
		r#"
let var xoo = "hello";
log(xoo);
    //^
"#
	);

	test_hover_list!(
		test_var_inside_preflight_closure,
		r#"
() => {
  let var goooo = "gar";
          //^
}"#
	);

	test_hover_list!(
		test_var_inside_inflight_closure,
		r#"
inflight () => {
  let var goooo = "gar";
          //^
}"#
	);

	test_hover_list!(
		test_builtin_instance_method,
		r#"
"hello".startsWith("h");
           //^
"#
	);

	test_hover_list!(
		multipart_reference_hover_middle,
		r#"
let j = Json {};
j.get("hello").get("world");
 //^
"#
	);

	test_hover_list!(
		map_element,
		r#"
let k = "hi";
{ k => "" }
//^
"#
	);

	test_hover_list!(
		json_element,
		r#"
{ hi: "cool" }
 //^
"#
	);

	test_hover_list!(
		json_element_nested_top,
		r#"
{ hi: { inner: [1, 2, 3] } }
 //^
"#
	);

	test_hover_list!(
		json_element_nested_inner,
		r#"
{ hi: { inner: [1, 2, 3] } }
        //^
"#
	);

	test_hover_list!(
		inflight_init,
		r#"
struct Data {
	field: str;
}

class T {
	new() {
		Data { field: "" };
	}

	inflight new() {
		Data { field: "" };
		//^
	}
}
"#
	);

	test_hover_list!(
		class_init_this_field,
		r#"
class T {
  stuff: num;

  new() {
    this.stuff = 1;
         //^
  }
}
"#
	);

	test_hover_list!(
		struct_decl_field,
		r#"
struct S {
  field: str;
  //^
}
"#
	);

	test_hover_list!(
		static_method_root,
		r#"
Json.stringify({});
//^
"#
	);

	test_hover_list!(
		variadic_args,
		r#"
  class Arr {
    pub addMany(...items: Array<str>) {
  
    }
  }
  
  let arr = new Arr();
  arr.addMany("a","b","c");
      //^
  "#,
	);
}
