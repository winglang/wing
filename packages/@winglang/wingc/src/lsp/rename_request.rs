use std::collections::HashMap;

use crate::lsp::sync::PROJECT_DATA;
use crate::visit::Visit;
use crate::wasm_util::extern_json_fn;
use lsp_types::{RenameParams, WorkspaceEdit};

use super::rename_visitor::RenameVisitor;
use super::sync::{check_utf8, WING_TYPES};

#[no_mangle]
pub unsafe extern "C" fn wingc_on_rename(ptr: u32, len: u32) -> u64 {
	extern_json_fn(ptr, len, on_rename_request)
}

pub fn on_rename_request(params: RenameParams) -> WorkspaceEdit {
	WING_TYPES.with(|types| {
		let types = types.borrow();
		PROJECT_DATA.with(|project_data| -> WorkspaceEdit {
			let project_data = project_data.borrow();
			let uri = params.text_document_position.text_document.uri;
			let file = check_utf8(uri.to_file_path().expect("LSP only works on real filesystems"));
			let scope = project_data.asts.get(&file).unwrap();

			let new_word = params.new_name;
			let position = params.text_document_position.position;

			let mut reference_visitor = RenameVisitor::new(&types);
			reference_visitor.visit_scope(scope);

			let text_edits = reference_visitor.create_text_edits(position, new_word.clone());

			let mut changes = HashMap::new();
			changes.insert(uri, text_edits);

			WorkspaceEdit {
				changes: Some(changes),
				document_changes: None,
				change_annotations: None,
			}
		})
	})
}

#[cfg(test)]
mod tests {
	use crate::lsp::rename_request::*;
	use crate::lsp::sync::test_utils::*;
	use lsp_types::Range;

	/// Creates a snapshot test for a given wing program's rename_request at a given position
	/// In the wing program, place a comment "//^" into the text where the "^" is pointing to the desired character position
	/// To assert on the target range of the result, place a comment "//-" below the target, with additional "-" characters to extend the range
	///
	/// First parameter will be the name of the tests, as well as the identifier to use for the list of completion in the asserts (see last parameter)
	/// Second parameter is the wing code block as a string literal
	/// Third one will be the new name to apply
	/// After the first two parameters, additional statements can optionally be provided to assert on the returned data.
	///
	/// Result is a Workspace Edit
	macro_rules! test_rename_request {
		($name:ident, $code:literal, $new_word:literal) => {
			test_rename_request!($name, $code, $new_word,);
		};
		($name:ident, $code:literal, $new_word:literal, $($assertion:stmt)*) => {
			#[test]
			fn $name() {
				// NOTE: this is needed for debugging to work regardless of where you run the test
				std::env::set_current_dir(env!("CARGO_MANIFEST_DIR")).unwrap();

				let text_document_position_params = load_file_with_contents($code);

				let $name = on_rename_request(RenameParams {
					text_document_position: text_document_position_params.clone(),
					work_done_progress_params: Default::default(),
          new_name: String::from($new_word)
				});

      let mut text_edit_ranges: Vec<Range> = vec![];
      if let Some(changes) = &$name.changes {
        for text_edits in changes.values() {
            for text_edit in text_edits {
              assert_eq!(
                $new_word,
                text_edit.new_text,
                "new_word did not match"
              );
              text_edit_ranges.push(text_edit.range.clone());
            }
        }
      }

        assert_eq!(
            get_ranges($code),
						text_edit_ranges,
						"target_range did not match"
					);

				$($assertion)*
			}
		};
	}

	test_rename_request!(
		variable_same_scope,
		r#"
let thing = "thing";
  //-----
let thing2 = thing;
           //----^
		"#,
		"t1"
	);

	test_rename_request!(
		variable_same_scope_declaration,
		r#"
let thing = "thing";
  //----^
let thing2 = thing;
           //-----
		"#,
		"t1"
	);

	test_rename_request!(
		rename_str,
		r#"
let thing = "thing";
               //^
let thing2 = thing;
		"#,
		"t1"
	);

	test_rename_request!(
		rename_symbol,
		r#"
let thing = "thing";
                 //^
let thing2 = thing;
		"#,
		"t1"
	);
	//

	test_rename_request!(
		rename_inner_scope,
		r#"
let var thing = 6;
      //-----
if (thing > 8) {
  //-----
  thing = 6;
//----^
} else {
  if (thing < 0) {
    //-----
    let thing = 0;
    log(thing);
  }
  log(thing)
    //-----
}               
		"#,
		"t1"
	);

	test_rename_request!(
		rename_outer_scope,
		r#"
    let var thing = 6;
          //----^
    if (thing > 8) {
      //-----
    thing = 6;
  //-----
    } else {
    if (thing < 0) {
      //-----
      let thing = 0;
      log(thing);
    }
    log(thing)
      //-----
    }
		"#,
		"t1"
	);

	test_rename_request!(
		rename_inner_scope_shadowing,
		r#"
    let var thing = 6;
    if (thing > 8) {
    thing = 6;

    } else {
    if (thing < 0) {
      let thing = 0;
        //-----
      log(thing);
        //----^
    }
    log(thing)
    }
		"#,
		"t1"
	);

	test_rename_request!(
		rename_class,
		r#"
class User {
    //----
  name: str;
}

class Admin extends User {
                  //----
  permission: num;
}

let a = new User();
          //---^
let b = new Admin();
		"#,
		"t1"
	);

	test_rename_request!(
		rename_closure,
		r#"
let pow = inflight (x: num) => {
  //--^
  return x * x;
};

test "pow" {
  assert(pow(5) == 25);
       //---
  assert(pow(25) == pow( pow(5) ));
       //---      //---//---
}
		"#,
		"t1"
	);

	test_rename_request!(
		rename_for_iterator,
		r#"
    bring cloud;

    let buckets = [new cloud.Bucket()];
    for bucket in buckets {
      //-----^
      bucket.addObject("bucket","bucket");
    //------
      log(bucket.toString());
        //------
    }
		"#,
		"t1"
	);

	test_rename_request!(
		rename_for_iterator_reference,
		r#"
    bring cloud;

    let buckets = [new cloud.Bucket()];
    for bucket in buckets {
      //------
      bucket.addObject("bucket","bucket");
    //------
      log(bucket.toString());
        //-----^
    }
		"#,
		"t1"
	);

	test_rename_request!(
		rename_for_loop,
		r#"
    bring cloud;

    let buckets = [new cloud.Bucket()];
      //------^
    for bucket in buckets {
                //-------
      bucket.addObject("bucket","bucket");
      log(bucket.toString());
    }
		"#,
		"t1"
	);

	// TODO: not supported yet
	test_rename_request!(
		rename_namespace,
		r#"
    bring cloud;
           //^
    let bucket = new cloud.Bucket();
		"#,
		"t1"
	);

	// TODO: not supported yet
	test_rename_request!(
		rename_namespace_from_property,
		r#"
    bring cloud;
    let bucket = new cloud.Bucket();
                     //^
		"#,
		"t1"
	);

	test_rename_request!(
		rename_struct_key,
		r#"
    struct User {
      name: str;
    //---^
    }
 
 let a: User = {name: "a"};
              //----

 log(a.name);
     //----
		"#,
		"t1"
	);

	test_rename_request!(
		rename_struct_key_usage,
		r#"
    struct User {
      name: str;
    //----
    }
 
 let a: User = {name: "a"};
              //----

 log(a.name);
     //---^
		"#,
		"t1"
	);

	test_rename_request!(
		rename_struct_name,
		r#"
    struct user {
         //----
      name: str;
    }
    
    let a: user = {name: "a"};
         //---^
		"#,
		"t1"
	);
	// TODO: not supported yet
	test_rename_request!(
		rename_interface_key,
		r#"
    interface ICat {
      inflight stepOnKeyboard(): str;
              //^
    }
    
    class Cat impl ICat {
      pub inflight stepOnKeyboard(): str {
                 
         return "/..,.o0e";
      }
    }
    
    let hershy = new Cat();
    
    test "cat misbehaves" {
      hershy.stepOnKeyboard();
    }
		"#,
		"t1"
	);

	test_rename_request!(
		rename_interface_name,
		r#"
    interface ICat {
            //---^
      inflight stepOnKeyboard(): str;
    }
    
    class Cat impl ICat {
                 //----
      pub inflight stepOnKeyboard(): str {
                 
         return "/..,.o0e";
      }
    }
    
    let hershy = new Cat();
    
    test "cat misbehaves" {
      hershy.stepOnKeyboard();
    }
		"#,
		"t1"
	);

	test_rename_request!(
		rename_class_property,
		r#"
class Cat {
  pub name: str;
    //---^
  new(name: str) {
    this.name = name;
       //----
  }

  inflight pub jump(height: num) {}
  pub sleep(d: duration) {
    log("{this.name} sleeps");
             //----
  }
}

let a = new Cat("A") as "a cat";
let b = new Cat("B") as "b cat";

a.sleep(18h);
b.sleep(24h);
log(b.name);
    //----

test "a cat scares b cat" {
  a.jump(5);
  b.jump(-5);
}
		"#,
		"t1"
	);
	test_rename_request!(
		rename_class_property_from_usage,
		r#"
class Cat {
  pub name: str;
    //----
  new(name: str) {
    this.name = name;
       //----
  }

  inflight pub jump(height: num) {}
  pub sleep(d: duration) {
    log("{this.name} sleeps");
             //----
  }
}

let a = new Cat("A") as "a cat";
let b = new Cat("B") as "b cat";

a.sleep(18h);
b.sleep(24h);
log(b.name);
    //---^

test "a cat scares b cat" {
  a.jump(5);
  b.jump(-5);
}
		"#,
		"t1"
	);

	test_rename_request!(
		rename_class_inflight_method,
		r#"
class Cat {
  pub name: str;
  new(name: str) {
    this.name = name;
  }

  inflight pub jump(height: num) {}
             //---^
  pub sleep(d: duration) {
    log("{this.name} sleeps");
  }
}

let a = new Cat("A") as "a cat";
let b = new Cat("B") as "b cat";

a.sleep(18h);
b.sleep(24h);
log(b.name);

test "a cat scares b cat" {
  a.jump(5);
  //----
  b.jump(-5);
  //----
}
		"#,
		"t1"
	);

	test_rename_request!(
		rename_class_inflight_method_from_usage,
		r#"
class Cat {
  pub name: str;
  new(name: str) {
    this.name = name;
  }

  inflight pub jump(height: num) {}
             //----
  pub sleep(d: duration) {
    log("{this.name} sleeps");
  }
}

let a = new Cat("A") as "a cat";
let b = new Cat("B") as "b cat";

a.sleep(18h);
b.sleep(24h);
log(b.name);

test "a cat scares b cat" {
  a.jump(5);
  //----
  b.jump(-5);
  //---^
}
		"#,
		"t1"
	);
	test_rename_request!(
		rename_class_preflight_method,
		r#"
class Cat {
  pub name: str;
  new(name: str) {
    this.name = name;
  }

  inflight pub jump(height: num) {}
  pub sleep(d: duration) {
    //----^
    log("{this.name} sleeps");
  }
}

let a = new Cat("A") as "a cat";
let b = new Cat("B") as "b cat";

a.sleep(18h);
//-----
b.sleep(24h);
//-----
log(b.name);

test "a cat scares b cat" {
  a.jump(5);
  b.jump(-5);
}
		"#,
		"t1"
	);

	test_rename_request!(
		rename_class_preflight_method_from_usage,
		r#"
class Cat {
  pub name: str;
  new(name: str) {
    this.name = name;
  }

  inflight pub jump(height: num) {}
  pub sleep(d: duration) {
    //-----
    log("{this.name} sleeps");
  }
}

let a = new Cat("A") as "a cat";
let b = new Cat("B") as "b cat";

a.sleep(18h);
//----^
b.sleep(24h);
//-----
log(b.name);

test "a cat scares b cat" {
  a.jump(5);
  b.jump(-5);
}
		"#,
		"t1"
	);

	test_rename_request!(
		new_scope_args,
		r#"
	  class Cat {
	    pub name: str;
	    new(name: str) {
	      //---^
	      this.name = name;
	                //----
	    }
	  }
	  "#,
		"t1"
	);

	test_rename_request!(
		named_args,
		r#"
	  struct A {
      option: str;
    //------
    }
    
    let foo = (opts: A) => {};
    
    let option = "myOption";
    foo(option: "bla");
      //-----^
    log(option);
	  "#,
		"t1"
	);

	test_rename_request!(
		named_args_renaming_struct,
		r#"
	  struct A {
      option: str;
    //-----^
    }
    
    let foo = (opts: A) => {};
    
    let option = "myOption";
    foo(option: "bla");
      //------
    log(option);
	  "#,
		"t1"
	);

	test_rename_request!(
		same_name_as_named_arg,
		r#"
    struct A {
      option: str;
    }
    
    let foo = (opts: A) => {};
    
    let option = "myOption";
      //------
    foo(option: "bla");
    log(option);
      //-----^
	  "#,
		"t1"
	);
}
