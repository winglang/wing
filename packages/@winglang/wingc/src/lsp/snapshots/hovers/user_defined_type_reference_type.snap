---
source: libs/wingc/src/lsp/hover.rs
---
contents:
  kind: markdown
  value: "```wing\nclass Foo {\n  // No public members\n}\n```"
range:
  start:
    line: 4
    character: 0
  end:
    line: 4
    character: 3

